### PR TITLE
Updates to testing

### LINKED ISSUE #631: Travis testing isn't working for Linux/OpenMPI and macOS
Travis testing should work for all cases. I made a new branch to work on this.

### PR DESCRIPTION
**Pull request by rieder**
- overhaul of travis config (for now only linux/mpich, but much easier to extend)
- pytest instead of nose
- move tests to amuse namespace
- some fixes to errors
- some fixes for deprecation warnings

Fixes #631, #414

Needs testing with #425, #351

### CODE REVIEW

**Comment by rieder** [comment]
I guess codecov is complaining since the whole test directory is gone now... that can be ignored.

**Comment by rieder** [comment]
@ipelupessy ok to merge?

File: src/amuse/io/nemobin.py [hunk 1]
```diff
@@ -273,147 +282,141 @@ def get_item_header(self):
...
-        
-    
+        self.file.write(x.tobytes())
+
     def write_array(self, typetag, data):
         x = array.array(typetag, data)
         x.append(0)
-        self.file.write(x.tostring())
-        
+        self.file.write(x.tobytes())
```
Comment (ipelupessy) [inline src/amuse/io/nemobin.py:334]: why is this?

Comment (rieder) [reply src/amuse/io/nemobin.py:334]: fixes a deprecation warning

File: src/amuse/support/methods.py [hunk 2]
```diff
@@ -1,24 +1,27 @@
 from amuse.support.core import late
 from amuse.support import exceptions
-from amuse.units import nbody_system
+# from amuse.units import nbody_system
```
Comment (ipelupessy) [inline src/amuse/support/methods.py:3]: if its not needed, remove?

File: src/amuse/io/nemobin.py (hunk 1 above)
Comment (rieder) [reply src/amuse/io/nemobin.py:334]: (or at least some sort of warning that came up... very specific in that it should be fixed this way.)

File: src/amuse/support/methods.py (hunk 2 above)
Comment (rieder) [reply src/amuse/support/methods.py:3]: ya

File: src/amuse/support/options.py [hunk 3]
```diff
@@ -90,137 +90,135 @@ def homedirectory(self):
...
     def override_value_for_option(self, name, value):
         self.overriden_options[name] = value
-        
 
     def read_from_ini_string(self, string):
         file = StringIO(string)
-        self.config.readfp(file)
-    
-    
+        self.config.read_file(file)
```
Comment (ipelupessy) [inline src/amuse/support/options.py:126]: confirm this is ok?

Comment (rieder) [reply src/amuse/support/options.py:126]: yes - again a deprecation warning with specific fix listed

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py [hunk 4]
```diff
@@ -597,6 +598,7 @@ def test30(self):
           "got multiple values for argument 'i' of method dummy_3_int")
         instance.stop()
 
+    @pytest.mark.skip
```
Comment (ipelupessy) [inline src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py:601]: hmmm. this test should work...

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import pytest
```
Comment (ipelupessy) [inline src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py:1]: I think its better to use the unittest skip

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py (hunk 4 above)
Comment (rieder) [reply src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py:601]: Yes - but it doesn't... same for test31 in fortran implementation.

File: src/amuse/test/suite/compile_tests/test_fortran_implementation.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import pytest
```
Comment (ipelupessy) [inline src/amuse/test/suite/compile_tests/test_fortran_implementation.py:1]: same as above

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py (hunk 5 above)
Comment (rieder) [reply src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py:1]: it's best not to skip but sure :)

File: src/amuse/test/suite/compile_tests/test_stopping_conditions.py [hunk 7]
```diff
@@ -573,7 +573,7 @@ def get_mpidir(cls):
         
     @classmethod
     def get_codestring(cls):
-        return CodeStringF
+        return codestringF
```
Comment (ipelupessy) [inline src/amuse/test/suite/compile_tests/test_stopping_conditions.py:576]: so is pytest still testing this class (it shouldn't)?

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py (hunk 4 above)
Comment (rieder) [reply src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py]: no idea why yet, but these tests hang.

File: src/amuse/test/suite/compile_tests/test_stopping_conditions.py (hunk 7 above)
Comment (rieder) [reply src/amuse/test/suite/compile_tests/test_stopping_conditions.py:576]: yes, this came up as an error.

File: src/amuse/test/suite/core_tests/test_attribute_storage.py [hunk 8]
```diff
@@ -10,55 +10,55 @@
...
 
 class _AbstractTestInMemoryAttributeStorage(amusetest.TestCase):
-    
+
     def new_inmemory_storage(self, is_with_units = True):
         raise NotImplementedError()
-        
-    def test1(self):
+
+    def xtest1(self):
```
[the remark this answers is not in the document]
Comment (rieder) [reply src/amuse/test/suite/core_tests/test_attribute_storage.py:17]: my bad - this was me disabling them temporarily. will revert.

File: src/amuse/test/suite/core_tests/test_store.py [hunk 9]
```diff
@@ -28,7 +29,8 @@ def store_version(self):
         
     def get_version_in_store(self, container):
         raise NotImplementedError
-        
+
+    @pytest.mark.skip
```
Comment (ipelupessy) [inline src/amuse/test/suite/core_tests/test_store.py:33]: don't think this should be the solution...

File: src/amuse/test/suite/core_tests/test_attribute_storage.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rieder) [reply src/amuse/test/suite/core_tests/test_attribute_storage.py:17]: https://docs.pytest.org/en/latest/goodpractices.html#conventions-for-python-test-discovery

File: src/amuse/test/suite/core_tests/test_store.py (hunk 9 above)
Comment (rieder) [reply src/amuse/test/suite/core_tests/test_store.py:33]: no

File: src/amuse/units/generic_unit_converter.py [hunk 10]
```diff
@@ -113,15 +113,15 @@ def matrixrank(self, A, tol=1e-8):
         return numpy.sum(numpy.where( s>tol, 1, 0 ) )
 
     def determine_new_base(self):
-        matrix = numpy.mat(numpy.zeros((self.system_rank,self.system_rank)))
+        matrix = numpy.asmatrix(numpy.zeros((self.system_rank,self.system_rank)))
```
Comment (ipelupessy) [inline src/amuse/units/generic_unit_converter.py:116]: I think pytest is warning that matrix is deprecated? should be ordinary array and then the matrix operations should be replaced ( * -> .dot( and **-1 -> numpy.linalg.inv())

Comment (rieder) [reply src/amuse/units/generic_unit_converter.py:116]: ya. sounds good.

Comment (rieder) [reply src/amuse/units/generic_unit_converter.py:116]: matrix is not yet deprecated but "pending deprecation" and not recommended. 
asmatrix and mat are the same, I found asmatrix easier to find when it needs to be replaced.

File: src/amuse/test/suite/core_tests/test_attribute_storage.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ipelupessy) [reply src/amuse/test/suite/core_tests/test_attribute_storage.py:17]: hmmm...https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure

File: src/amuse/test/suite/compile_tests/test_stopping_conditions.py (hunk 7 above)
Comment (ipelupessy) [reply src/amuse/test/suite/compile_tests/test_stopping_conditions.py:576]: the change is ok, but it shouldn't test the class

File: src/amuse/test/suite/core_tests/test_store.py (hunk 9 above)
Comment (rieder) [reply src/amuse/test/suite/core_tests/test_store.py:33]: how about adding an is_abstract method to the class, skip if it returns True?

Comment (rieder) [reply src/amuse/test/suite/core_tests/test_store.py:33]: seems clumsy but I can't figure out a better method...

Comment (rieder) [reply src/amuse/test/suite/core_tests/test_store.py:33]: never mind, can just add @unittest.skip() to a class

Comment (rieder) [reply src/amuse/test/suite/core_tests/test_store.py:33]: but then that gets inherited... brrrr....

**Comment by ipelupessy** [comment]
I am still testing and running into issues (which may be my config)..

**Comment by rieder** [comment]
ok - travis runs ok apart from the issue with test_attribute_storage's setup class not being skipped...

**Comment by rieder** [comment]
(and the test31 issues)

**Comment by ipelupessy** [comment]
the last commit fixes the case where baseclass is tested (when it shouldn't). Apperently pytest also tries all classes derived from UnitTestCase.. (not sure if that is documented). Can be solved by making it mixin

**Comment by ipelupessy** [comment]
so you can also fix the other cases this way...

File: src/amuse/units/generic_unit_converter.py (hunk 10 above)
Comment (rieder) [reply src/amuse/units/generic_unit_converter.py:116]: this is not immediately necessary so I'll leave it for now

**Comment by ipelupessy** [comment]
have you tested the codes test?

**Comment by ipelupessy** [comment]
there are some codes test which leave debris in the cwd - capreole and and maybe others..setup and teardown with chdir can be added for those...

**Comment by rieder** [comment]
@ipelupessy That timeout you removed was a timeout override, allowing more time for that specific tests... I guess we could allow more time in general but I'd rather not do that and keep the override instead.

File: src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py (hunk 4 above)
Comment (rieder) [reply src/amuse/test/suite/compile_tests/test_c_implementation_simplified.py]: it should, but it doesn't. test31 in test_c_implementation_simplified.py and test_fortran_implementation is problematic for some reason, both on travis and for me locally (macOS + Linux). I don't know why...

**Comment by ipelupessy** [comment]
ok, I had some issues with a pytest not supporting this...

**Comment by rieder** [comment]
yes, it's a plugin for pytest (pip install pytest-timeout)

**Comment by rieder** [comment]
but I'm just allowing more time for core-tests now, should be ok (these don't take long generally)

**Comment by rieder** [comment]
I'll merge if you're ok with it. Some tests need more addressing, but that's the same for the old testing setup...

**Comment by rieder** [comment]
@ipelupessy ok?

**Comment by ipelupessy** [comment]
ok

**Pull request merged**